src/compiler/crystal/codegen/unions.cr

```diff
@@ -41,6 +41,10 @@ module Crystal
   end
 
   class CodeGenVisitor
+    def union_type_and_value_pointer(union_pointer)
+      {load(union_type_id(union_pointer)), union_value(union_pointer)}
```

Comment:
What do you think if we define `union_type_id` and `union_value` as local variables here:

```crystal
def union_type_and_value_pointer(union_pointer)
  union_type_id = aggregate_index union_pointer, 0
  union_value = aggregate_index union_pointer, 1
  {load(union_type_id, union_value)}
end
```

Then we can redefine `def union_value` and `def union_type_id`:

```crystal
def union_type_id(ptr)
  union_type_and_value_pointer(ptr)[0]
end

def union_value(ptr)
  union_type_and_value_pointer(ptr)[1]
end
```

Then you can still use `union_value(...)` and `union_type_id(...)` instead of using `_, value = union_type_and_value_pointer(...)` in many places.

And the knowledge of how a union is stored is still kept in a single place: `union_type_and_value_pointer`.

Another nice thing about this is that we could eventually "optimize" the two methods to avoid generating LLVM code that's not used (I don't think the extra code will appear in the final code, but it's still invoking some LLVM apis).

Reply:
I guess there is `)` missplaced:

```cr
def union_type_and_value_pointer(union_pointer)
  union_type_id = aggregate_index union_pointer, 0
  union_value = aggregate_index union_pointer, 1
  {load(union_type_id), union_value}
end
```

The llvm values are not emitted unless it's used I think. Or they should be removed since they are not used.

I understand that having those 2 methods seems easier to read, but will allow misguided use of unions in a MT environment. The value could change between the two operations. Alternative when reading a value the supposed type must be passed and that API is weirder.

I would even remove (or make private) the `union_type_id` and `union_value` methods.

Reply:
> but will allow misguided use of unions in a MT environment

How is this possible if the methods only read?

I mean, how is this:

```crystal
_, value = union_type_and_value_pointer(ptr, type)
```

safer than this:

```crystal
value = union_value(ptr, type)
```

given that `union_value(ptr, type)` is just `union_type_and_value_pointer(ptr, type)[1]`?

Reply:
I'm only proposing `union_value(ptr, type)` to be a shorter way to write `_, value = union_type_and_value_pointer(ptr, type)`.

Reply:
It won't be wrong by itself, but later one will use `union_type_id(ptr)` and `union_value(ptr)` independently in the same codegen method and miss the point that type and value should be read in the same operation.

Reply:
Ooooh... I see what you mean.

Then yes, that makes sense.

Should we remove the methods `union_value` and `union_type_id` then, and inline them in the other one?

Reply:
On second thoughts, the union_type_id and union_value have some occurrences here. We could either leave them as private or even more accurate, file's private. Since the codegen is a heavily used partial class, making it instance private method will not prevent its usage.

Is there any objection to converting them into private file methods?

Reply:
I see `union_type_and_value_pointer` as the union read operation.
`store*_in_union` as the writer operations.

And `union_type_id` and `union_value` as helpers for the particular representation of the unions. That should be used in the above methods. I don't want to replace `union_type_id` for a call to `union_type_and_value_pointer`.

Reply:
Sounds good. File private or not, it's fine to leave the helpers 👍

Reply:
Ok, let's merge it as is. Making it file private requires passing the codegen as argument, the refactor can be done later. There are other implementations helper methods in the codegen.